### PR TITLE
Phase 1 runtime composition

### PR DESCRIPTION
**Pull request by SuperCoolPencil**
<h3>Greptile Summary</h3>

This PR introduces the Phase 1 runtime composition layer: a new `internal/runtime.App` struct that owns the pool/service/lifecycle wiring, a `cmd/runtime_state.go` compatibility shim that keeps the existing `Global*` mirrors in sync during the transition, and a `architecture.md` document describing the four-layer target architecture. The overall direction is sound and the `internal/runtime` package itself is well-structured with correct `RLock`/`Lock` usage and an idempotent hook-wiring guard.

Key concerns:

- **Service teardown race** (`cmd/runtime_state.go:173`): The three-step pattern `currentApp()` → `EnsureLocalServiceAndLifecycle()` → `syncLegacyGlobalsFromApp()` has a window where `a.service` is set internally but `GlobalService` is still `nil`. A concurrent `currentApp()` call in this window will classify the divergence as "out of sync", create a new app, and shut down the just-initialized service. This is a new regression relative to the old code.
- **Unguarded write to `globalApp`** (`cmd/shutdown.go:36`): `resetGlobalShutdownCoordinatorForTest` sets `globalApp = nil` without holding `legacyLifecycleMu`, which is inconsistent with every other mutation of that variable and would be a data race if any test uses `t.Parallel()`.
- **Stale `lifecycle` capture in hook wiring** (`internal/runtime/app.go:212`): `lifecycle` is captured from `ensureLocalLifecycle` (under `lifecycleMu`) and then used later inside a separate `componentsMu` section. A concurrent `ApplyComponents()` call between these two sections can replace `a.lifecycle`, leaving the app with an unwired lifecycle while hooks are attached to the detached one.
- **Missing test coverage**: `initLocalRuntime` and the `globalApp == nil` first-call path in `currentApp()` have no tests.

<h3>Confidence Score: 2/5</h3>

- Not safe to merge as-is; the race window in `ensureGlobalLocalServiceAndLifecycle` can silently shut down a just-initialized service under concurrent load, and the unguarded write to `globalApp` in the test reset helper is inconsistent with the mutex discipline established elsewhere.
- The `internal/runtime` package is well-implemented and the architecture document is clear, but the `cmd` compatibility shim introduces at least two correctness issues: (1) a race window where a concurrent `currentApp()` call can interpret the transient divergence between `a.service` and `GlobalService` as an "out of sync" signal and tear down a newly created service; and (2) a direct unsynchronized write to `globalApp` in a test helper. Both are byproducts of the design where a global-mirror comparison is used to detect divergence, but the mirror is updated in a separate step after the mutation.
- `cmd/runtime_state.go` (race window around `ensureGlobalLocalServiceAndLifecycle`) and `cmd/shutdown.go` (unguarded `globalApp = nil` in the test reset helper) need the most attention before merging.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| architecture.md | New documentation-only file describing the four-layer target architecture; no logic or code concerns. |
| cmd/root.go | Removes per-global initialization boilerplate in favour of `initLocalRuntime`; `publishSystemLog` still reads `GlobalService` without holding `legacyLifecycleMu` but this is a pre-existing pattern in the compatibility layer. |
| cmd/runtime_state.go | Core Phase 1 compatibility shim; contains a race window in the currentApp()→EnsureLocalServiceAndLifecycle()→syncLegacyGlobalsFromApp() sequence that can cause a just-initialized service to be torn down by a concurrent caller. |
| cmd/runtime_state_test.go | Good coverage of the concurrent app-replacement and non-blocking-shutdown paths; missing coverage for first-call (globalApp == nil) and the lifecycle-stale-detection branch. |
| cmd/shutdown.go | Simplifies shutdown to a single App.Shutdown() call; `resetGlobalShutdownCoordinatorForTest` writes `globalApp = nil` directly without holding `legacyLifecycleMu`, introducing an unsynchronized access inconsistent with the rest of the codebase. |
| internal/runtime/app.go | New runtime App struct with correct RLock/Lock usage and idempotent hook wiring; the lifecycle variable captured in EnsureLocalServiceAndLifecycle can become stale if ApplyComponents is called concurrently between the two componentsMu sections. |
| internal/runtime/app_test.go | Solid tests for lifecycle reset, NewLocal initialization, concurrent-init safety, and idempotent hook wiring; concurrent test should be run with -race to be meaningful. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant cmd as cmd caller
    participant CS as currentApp()
    participant App as runtime.App
    participant Globals as Global* mirrors
    participant Sync as syncLegacyGlobalsFromApp()

    Note over cmd,Globals: Normal happy-path (single goroutine)
    cmd->>CS: ensureGlobalLocalServiceAndLifecycle()
    CS->>Globals: acquire legacyLifecycleMu
    CS->>App: executionComponentsOutOfSyncLocked() → false
    CS->>App: ApplyComponents(current globals)
    CS->>Globals: release legacyLifecycleMu
    CS-->>cmd: returns App

    cmd->>App: EnsureLocalServiceAndLifecycle()
    App->>App: componentsMu.Lock() → create S1 → a.service = S1
    App->>App: ensureLocalLifecycle() → create lifecycle
    App->>App: wire engine hooks
    App-->>cmd: nil (success)

    Note over App,Globals: ⚠️ Race window: a.service=S1 but GlobalService=nil
    cmd->>Sync: syncLegacyGlobalsFromApp()
    Sync->>Globals: acquire legacyLifecycleMu
    Sync->>App: Components() → {S1, ...}
    Sync->>Globals: GlobalService = S1
    Sync->>Globals: release legacyLifecycleMu

    Note over cmd,Globals: Concurrent caller during race window
    participant G2 as goroutine 2
    G2->>CS: currentApp()
    CS->>Globals: acquire legacyLifecycleMu
    CS->>App: executionComponentsOutOfSyncLocked()
    App-->>CS: S1 != nil (GlobalService) → TRUE ‼️
    CS->>App: appToShutdown = globalApp
    CS->>App: globalApp = NewEmpty()
    CS->>Globals: release legacyLifecycleMu
    CS->>App: appToShutdown.Shutdown() ← shuts down S1 ‼️
    CS-->>G2: returns empty replacement app
```

<details open><summary><h3>Comments Outside Diff (3)</h3></summary>

1. `internal/runtime/app_test.go`, line 1-46 ([link](https://github.com/surge-downloader/surge/blob/b2d4f299105e6678923e8200d5ab270ff744e97d/internal/runtime/app_test.go#L1-L46)) 

   **Critical integration paths are untested**

   The two tests only verify happy-path construction. The most complex and failure-prone code paths are completely uncovered:

   - `EnsureLocalServiceAndLifecycle` — the core wiring function that initialises the service, lifecycle, and engine hooks. No test exercises it at all, let alone edge cases like concurrent calls or calling it when `pool` is nil.
   - `LifecycleForService` — the conditional logic (`service == nil || a.service == nil || service != a.service`) has no test.
   - `Shutdown` idempotency — `shutdownOnce` guarantees a single execution, but there is no test confirming double-`Shutdown()` is safe and returns the same error.
   - `ApplyComponents` / `Components` round-trip — no test verifies components written via `ApplyComponents` are read back correctly.
   - `CancelEnqueue` after context expiry — no test covers calling `CancelEnqueue` on a context that was already cancelled by `ResetEnqueueContext`.

   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   <details><summary>Prompt To Fix With AI</summary>
   
   `````markdown
   This is a comment left during a code review.
   Path: internal/runtime/app_test.go
   Line: 1-46
   
   Comment:
   **Critical integration paths are untested**
   
   The two tests only verify happy-path construction. The most complex and failure-prone code paths are completely uncovered:
   
   - `EnsureLocalServiceAndLifecycle` — the core wiring function that initialises the service, lifecycle, and engine hooks. No test exercises it at all, let alone edge cases like concurrent calls or calling it when `pool` is nil.
   - `LifecycleForService` — the conditional logic (`service == nil || a.service == nil || service != a.service`) has no test.
   - `Shutdown` idempotency — `shutdownOnce` guarantees a single execution, but there is no test confirming double-`Shutdown()` is safe and returns the same error.
   - `ApplyComponents` / `Components` round-trip — no test verifies components written via `ApplyComponents` are read back correctly.
   - `CancelEnqueue` after context expiry — no test covers calling `CancelEnqueue` on a context that was already cancelled by `ResetEnqueueContext`.
   
   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `cmd/root.go`, line 585-593 ([link](https://github.com/surge-downloader/surge/blob/ada8e8ec83ef70d3e754e4eb7c8578e1d84ce081/cmd/root.go#L585-L593)) 

   **`GlobalPool` nil dereference will panic**

   `GlobalPool` is accessed inside the `activeDownloadsFunc` closure without a nil guard. Because `GlobalPool` is a package-level variable that is updated asynchronously via `syncLegacyGlobalsFromApp`, there is a window during startup (or in any test scenario that does not call `initLocalRuntime`) where `GlobalPool` is `nil`. Calling `GlobalPool.GetAll()` on a nil pointer panics.

   The handler already guards against a nil `service` check above (line 563), but no equivalent guard exists for `GlobalPool`. If the pool is nil, `CheckForDuplicate` should receive an empty map rather than a panic:

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: cmd/root.go
   Line: 585-593

   Comment:
   **`GlobalPool` nil dereference will panic**

   `GlobalPool` is accessed inside the `activeDownloadsFunc` closure without a nil guard. Because `GlobalPool` is a package-level variable that is updated asynchronously via `syncLegacyGlobalsFromApp`, there is a window during startup (or in any test scenario that does not call `initLocalRuntime`) where `GlobalPool` is `nil`. Calling `GlobalPool.GetAll()` on a nil pointer panics.

   The handler already guards against a nil `service` check above (line 563), but no equivalent guard exists for `GlobalPool`. If the pool is nil, `CheckForDuplicate` should receive an empty map rather than a panic:

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

3. `cmd/root.go`, line 172 ([link](https://github.com/surge-downloader/surge/blob/ada8e8ec83ef70d3e754e4eb7c8578e1d84ce081/cmd/root.go#L172)) 

   **Stale `service` reference vs live `GlobalPool` creates inconsistent duplicate detection**

   `GlobalService` is captured by value at server startup and passed into every HTTP handler closure as `service`. If `currentApp()` later replaces the runtime app (because `executionComponentsOutOfSync()` returns true) and `syncLegacyGlobalsFromApp()` updates `GlobalPool`, the `activeDownloadsFunc` in `handleDownload` (line 585-592) reads the **new** pool, while the `service` parameter still refers to the **old** service. This leads to a split-brain state:

   - Duplicate detection at line 593 interrogates the **new** pool's in-flight downloads.
   - `lifecycle.Enqueue` / `service.Add` at lines 658-668 submits the download to the **old** service (because `lifecycleForLocalService(service)` observes `service != currentService` and may return `nil`, falling back to `service.Add` on the stale reference).

   A download could silently disappear into the old, already-shut-down service. Accessing `GlobalPool` live from within a handler bound to a specific `service` instance bypasses the ownership model that `LifecycleForService` is trying to enforce. The handler should derive pool data from `service` directly (or from `currentApp()`) rather than reading the global.

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: cmd/root.go
   Line: 172

   Comment:
   **Stale `service` reference vs live `GlobalPool` creates inconsistent duplicate detection**

   `GlobalService` is captured by value at server startup and passed into every HTTP handler closure as `service`. If `currentApp()` later replaces the runtime app (because `executionComponentsOutOfSync()` returns true) and `syncLegacyGlobalsFromApp()` updates `GlobalPool`, the `activeDownloadsFunc` in `handleDownload` (line 585-592) reads the **new** pool, while the `service` parameter still refers to the **old** service. This leads to a split-brain state:

   - Duplicate detection at line 593 interrogates the **new** pool's in-flight downloads.
   - `lifecycle.Enqueue` / `service.Add` at lines 658-668 submits the download to the **old** service (because `lifecycleForLocalService(service)` observes `service != currentService` and may return `nil`, falling back to `service.Add` on the stale reference).

   A download could silently disappear into the old, already-shut-down service. Accessing `GlobalPool` live from within a handler bound to a specific `service` instance bypasses the ownership model that `LifecycleForService` is trying to enforce. The handler should derive pool data from `service` directly (or from `currentApp()`) rather than reading the global.

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<sub>Last reviewed commit: 8d81579</sub>

> Greptile also left **4 inline comments** on this PR.

**Context used:**

- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

### CODE REVIEW

File: cmd/runtime_state.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	}
+	if GlobalLifecycleCleanup == nil {
+		cleanup, err := startLifecycleEventWorker(service, GlobalLifecycle)
+		if err != nil {
+			return nil, err
+		}
+		GlobalLifecycleCleanup = cleanup
+	}
+	return GlobalLifecycle, nil
+}
```
[the remark this answers is not in the document]
Comment (SuperCoolPencil) [reply cmd/runtime_state.go:263]: this is a work in progress architecture change, I am gonna create a bunch of stacked PRs, this will be removed in the future

**Comment by SuperCoolPencil** [comment]
* **#226** <a href="https://app.graphite.com/github/pr/surge-downloader/Surge/226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#225** <a href="https://app.graphite.com/github/pr/surge-downloader/Surge/225?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#223** <a href="https://app.graphite.com/github/pr/surge-downloader/Surge/223?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/surge-downloader/Surge/223?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
